scripts/staging/fm-binclass.dml

```diff
@@ -0,0 +1,179 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+
+/*
+ * Factorization Machines for binary classification.
+ */
+
+# Imports
+source("nn/optim/adam.dml") as adam
+source("nn/layers/fm.dml") as fm
+source("nn/layers/log_loss.dml") as log_loss
+source("nn/layers/sigmoid.dml") as sigmoid
+source("nn/layers/l2_reg.dml") as l2_reg
+source("nn/layers/cross_entropy_loss.dml") as cross_entropy_loss
+
+train = function(matrix[double] X, matrix[double] y, matrix[double] X_val, matrix[double] y_val)
+    return (matrix[double] w0, matrix[double] W, matrix[double] V, double loss) {
+  /*
+   * Trains the FM model.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - y : label corresponds to n examples
+   *  - lambda : regularization (5e-04)
+   *
+   * Outputs:
+   *  - w0, W, V : updated model parameters.
+   *  - loss : computed loss with log_loss.
+   *
+   * input propagation through layers
+   * fm::init -> adam::init -> fm::forward -> sigmoid::forward -> log_loss::forward \
+   *          adam::update <- fm::backward <- sigmoid::backward <- log_loss::backward <-
+   */
+
+    n = nrow(X);
+    d = ncol(X);
+    k = 2; # factorization dimensionality, only(=2) possible for now.
+ 
+    # 1.initialize fm core
+    [w0, W, V] = fm::init(n, d, k);
+    
+    # 2.initialize adam optimizer
+    ## Default values for some parameters
+    lr      = 0.001; 
+    beta1   = 0.9;       # [0, 1)
+    beta2   = 0.999;     # [0, 1)
+    epsilon = 0.00000001; 
+    t       = 0;
+    
+    # [mX, vX] = adam::init(X); # to optimize input.
+    [mw0, vw0] = adam::init(w0);
+    [mW, vW]   = adam::init(W);
+    [mV, vV]   = adam::init(V);
+    
+    # Regularization
+    lambda = 5e-04
+
+    # Optimize
+    print("Starting optimization")
+    batch_size = 10
+    iters = ceil(1000 / batch_size)
+    epochs = 100; N = 1000
+    for (e in 1:epochs) {
+      for (i in 1:iters) {
+        # Get the next batch
+        beg = ((i-1) * batch_size) %% N + 1
+        end = min(N, beg + batch_size - 1)
+        X_batch = X[beg:end,]
+        y_batch = y[beg:end,]
+    
+        # 3.Send inputs through fm::forward
+        y_res = fm::forward(X_batch, w0, W, V);
+    
+        # 4.Send the above result through sigmoid::forward
+        sfy = sigmoid::forward(y_res);
+    
+        # 5.Send the above result through log_loss::forward
+        loss = log_loss::forward(sfy, y_batch);
+
+        # Compute loss & accuracy for training & validation data every 100 iterations.
+        if (i %% 100 == 0) {
+          # Compute training loss & accuracy
+          loss_data = log_loss::forward(sfy, y_batch);
+          loss_reg_X  = l2_reg::forward(X, lambda);
```

Comment:
Please don't regularize the inputs `X`.  We only want to regularize (place priors upon) the parameters of our model.